### ISSUE TITLE
Release 1.5 discussion

### ISSUE THREAD

**Issue by jzoss** [jzoss/Git-Source-Control-Provider#17]
Place to talk about the nightly build

**Comment by jzoss** [comment]
OK.. so.. I am starting to think the CI build server might not be exactly right.. I get a different size build and It seems to have some bugs that are not present when I build locally. Anyway, I made a release to here https://github.com/jzoss/Git-Source-Control-Provider/releases/tag/1.4.2.36 .. it should be the same but it's not.. 

Anyway.. Please check out the new build and test it.. It should be fast.. I think it is faster than 1.4.2 .. If you are really bored.. install the CI build first.. check it out.. then the release from github..

**Comment by PureKrome** [comment]
Pew^Pew

**Comment by PureKrome** [comment]
Still getting VS crashes :/

I _think_ it's when i switch a branch (using TGit from Explorer) to an existing branch .. which has some commits already outstanding.

and then i open the sln (VS is already open) and :hankey: happens.

**Comment by jzoss** [comment]
Blerg.. OK.. Is that with the c.I. build or the one I posted in the release tab? Anyway.. At least I should be able to recreate that..

**Comment by PureKrome** [comment]
*Release Tab* build.

**Comment by jzoss** [comment]
Ok.. is there anything in the activity log... I can recreate it yet.. I have a few ideas..

**Comment by jzoss** [comment]
@PureKrome .. OK.. I did some digging.. who would have thought converting from an infinite loop to asynchronous events would be so difficult :) .. anyway.. I have a new [build](https://ci.appveyor.com/api/buildjobs/8ko8t6mh25mlhu58/artifacts/GitSccProvider/bin/Release/GitSccProvider.vsix) I think it's better.. Anyway before the crashing and stuff.. how was the performance?

**Comment by PureKrome** [comment]
>is there anything in the activity log.

where do i find this?

**Comment by jzoss** [comment]
%APPDATA%\Microsoft\VisualStudio\Version\ActivityLog.xml. ... But now that I've read about it.. I'm guessing I failed to log to it correctly

**Comment by teebee76** [comment]
Just got the latest CI-build, and right off I still notice the bug with the .gitignore Icon is still seperated from the .gitignore "menu" 
Illustrated by this image.
![image](https://cloud.githubusercontent.com/assets/5390116/11894229/1ce4ba0a-a576-11e5-9bcf-1db1bd2424b4.png)

Next I think maybe I figured out why my pending changes window wasn't displaying the changes (reported in #16) 
I noticed that when I change a file a few seconds later the glyph would change, but nothing would appeared in the pending changes window.

I then noticed that the pending Changes window would show pending changes for my master branch, even though I am currently in mye develop-branch. 
(which I see if I open the Git Bash window)

I then tried switching branch, but nothing seems to happen. And if I then try to do other stuff in VS I notice it has become very slow and laggy  and then it becomes alltogether unresponsive.  And finally crashes.. 

I will check out your build published here under releases and see if the non-CI-version is behaving diffrently

**Comment by teebee76** [comment]
Testing the build  you linked to in your post 6 hours back, I notice the same behaviour.

I just tried changing branch from the pending changes window (still not showing the proper branch) I first had to wait about 30 seconds for the window to dissappear, and then another 15 seconds before I got the "file open in editor changed from outside source, update or ignore.." 
the editor is pretty slow after the change..  just clicking the tab to view another open file for instance takes about 2 seconds. It used to be instantanious. 
If I check with the git bash window, the right branch is checked out, however Pending Changes remains at master, no matter what branch I switch to.

**Comment by jzoss** [comment]
@PureKrome , @teebee76 --- So.. I stayed up way to late last night.. and spent more time today than I should :) working on a new build. Anyway.. I did not fix the button on the gitignore yet :) Buttons are strangely complicated in VS extensions... Anyway.. checkout the new C.I. build.. I know the branch name is fixed.. I think the performance slow down is fixed now too.. but you never know.. it could show up again.. 

I'm thinking of adding a new feature.. not sure what you think.. maybe @teebee76 would care more.. "Track selected Repo in Pending changes.. so basiclly what ever file you select in the solution explorer it switch pending changes to what ever repo that is in.. 

Anyway my brain hurts now.. and the weekend is coming .. I don't get to see Star Wars till Sunday, so I may have time to work on it before hand..

**Comment by PureKrome** [comment]
Sometimes finding it slow for the Pending Changes file list, to actually realise i have some new files that are changed.

**Comment by PureKrome** [comment]
Totally understand that it's holidays, etc: Just recording my experience so far with `1.4.45` - still no files listed under `Pending Changes`.

**Comment by jzoss** [comment]
That makes me sad :( .. Pending changes I thought was working for me.. I'm technically on holiday till next week.. But if my wife let's me I may play a bit tonight

**Comment by PureKrome** [comment]
It's all good - RL/Holidays/Wife > OSS projects.

**Comment by teebee76** [comment]
Sorry for staying so quite for so long, but I had to work on other non c# projects, and while doing a hotfix on one of the c# projects I had to disable the plugin and use sourcetree instead, because the plugin was slowing VS down. still with the 1.4.49 release, I will check more in start of january or maybe tomorrow when I will spend some hours at work. At home with the kid today, just enough time while he's sleeping to log in and write this comment before doing some house cleaning (yuck)

@jzoss  did you get to see Star Wars by the way, and did you like it. No spoilers please, because I have yet to see it :) 

wish you guys a happy holliday and a fantastic, productive and bug-free new year  :smile:

**Comment by jzoss** [comment]
@teebee76 no worries :) first star wars was good... Not Empire Strikes Back good.. But better than all the prequels.

So this release is a big clean up.. I rewrote all the git integration stuff trying to make it fast and use less memory. Anyway I'm want to get through this and make it awesome fast. But the big reason I wanted to fix it all was so I could start adding new features.. And truth is, for all you people who have been so helpful and motivating .. I am very inclined to add features :) I just want to make this the best git plugin for visual studio cause I'm a but crazy.

**Comment by jzoss** [comment]
@teebee76 Question for you.. can you describe your visual studio project? I'm trying to track down the slowness.. Does it have multiple repos? How many files-ish? Just trying to reproduce that error :) Also if you could ever give me read only access to the code.. So I could debug.. that might help.. but I don't expect that at all :)

**Comment by teebee76** [comment]
sure. I think read access will be difficult, I have the repo on a "local" remote server, but I can summarize the specs for the solution.

The solution contains 3 projects and they are all under 1 git repository.

The folder structure is as follows
\solution\.git
\solution\project1
\solution\project2
\solution\project3   
\solution\packages (Nuget)

including all binaries and object folders the solution amounts to shockingly 176mb containing 5836 files and 561 folders...

delwing into the projects and excluding the obj and bins reveals the following

proj1: (main project) 
  452 files  
  38 folders 
  9,7mb
proj2: 
  22 files
  2 folders
  49kb
proj3:
  20 files
  2 folders
  59 kb

another thing that strikes me a bit odd when looking in the solutions folder is the size of the solution-user-options file ... it's 13mb in size .. don't really know what it contains and what is the usual size for this kind of file, but if it's only preference settings and the like it shouldn't have to be 13million bytes :)

I will take my computer home today, and log in later to do some commits and changes to look at the performance.

I will let you know if it is still an issue, I will also check to see if there is a newer nightly build I can try out.

**Comment by jzoss** [comment]
So I had a bit of time to try some thing today.. I THINK.... That I have a good chance that fixed a big performance issue. @teebee76 You should test it and tell me if you see a difference .. 


@PureKrome .. I added a new feature that may fix the Pending changes window. Check it out in the options. 

![capture](https://cloud.githubusercontent.com/assets/3586254/12103528/798ad51a-b308-11e5-8a12-4f122ab3f969.PNG)

I have to think of a better name... but it escapes me right now.. anyway.. it makes sure the pending changes window is always on the repo to the of the currently edited file... So if you have more than one repo, it switches for you basically.

**Comment by PureKrome** [comment]
>So if you have more than one repo, it switches for you basically.

I don't understand this statement.

When I open VS doesn't it just look for the `.git` folder that should be in the root folder, for the solution?

so if i keep opening up multiple VS's, then each VS instance is referring to a single `.git` folder / active repo.

Which means having that checkbox option, is not required?

**Comment by jzoss** [comment]
Sorry... I was really not clear.. I meant to say .. if you have multiple git repo under 1 solution.. I thought you said you had it that way.... but I was totally wrong :) All this does is IF you do have multiple repos.. like a submodule or whatever it handles that.

**Comment by PureKrome** [comment]
AHHH! i've never had multiple repo's per solution.

i never knew you could, either :confused: 

I'll also never do that, too :)

**Comment by jzoss** [comment]
@PureKrome I also never have done it myself... but I guess some people do.. Sorta a big pain.. I guess if you have a submodule.. Which I sorta hate... Anyway.. I found 2 big bugs.. that could be causing the slow downs.. I can't break pending changes myself... yet... but tell me if it's still not working.

**Comment by PureKrome** [comment]
new release is up, somewhere?

**Comment by jzoss** [comment]
@PureKrome Yea... the CI build.. There is a sorta big-ish bug.. Switching can crash Visual Studio.. But I think I just about have it fixed.

**Comment by jzoss** [comment]
So there is a new C.I. Build.. for some reason the version updating got messed up .. OK I fixed a few more bugs... not there yet but I feel I am actually close to a stable release.. But I think this one is clear of most bugs.. Should faster too.. Reshaper not causing issues yet.. Anyway.. I know the branch name is not updating correctly just yet... The current C.I. build is 1.4.52...  bed time..

**Comment by teebee76** [comment]
first sorry for not having time to look at this yesterday evening, I got to the computer too late after beeing forced to do some house-cleaning instead. And when I was about to release the new feature I noticed that the external consultant made an error that caused sql-issues to the system, so a bit stressed about that, it beeing a production system and all .. anyways ... 

just wanted to say, that in my other solution I have several git-repos, actually one for each project (at least the ones I want source-controlled. Small ones I don't bother.. Benefits as I see it is that you can have easily have CI for each project and have them publish seperately ... The solution is a bunch of console-tools that do various stuff, and some depend on each other, so when in one solution I don't have to bother opening several VS to do changes to the libraries as well as the console apps. .. At least this was before I started building nuget-packages of the libraries to my local nuget-folder and doing package-update... 

Although I am still "lazy", so I like having them in one solution to avoid filling my screens with different Visual Studio windows :)

I will (after the darned consultant fixes the error they made) have time to get the latest version and see if it improves the performance, I will also open my libraries/console-app-solution to see how that checkbox works.

**Comment by teebee76** [comment]
yeah, and @jzoss , is there a way to make that CI-link sticky or something? I always have a hard time finding it :D

**Comment by jzoss** [comment]
Yea.. It's in the read me on the front page.. But I'll find a better way later.. Anyway check out the new feature then.. You may like it :) OK very late.. This release had turned out to be much more complicated than I thought..

**Comment by teebee76** [comment]
I'm back at work after about 4 hours of sleep I guess, and my son was crying one of them... I hope you will sleep better, and keep up the good work :+1:  I will look for the link again :)

**Comment by jzoss** [comment]
Yes.. My 3.5 year old boy and 2 year old girl both have a cold... At least they are sleeping now.. I love when contractors give your more work.. Don't you :)

**Comment by jzoss** [comment]
Hey guy.. .got a bit more time to work on it.. New C.I. build is up (Just for you @teebee76 https://goo.gl/dF48Tp ) Anyway.. Threading is moving along. Starting to move faster.. It uses a lot less memory now.. I need to clean up a few more little things.. like the .gitignore bug.. and I want to clean up the filewatcher just a bit.. BUT.. I think I am close to a beta 1 build..

**Comment by PureKrome** [comment]
Nice work :) keep up the effort!

**Comment by teebee76** [comment]
Been sick a couple of days  previous release i checked out worked a øot better for me,  some things i noticed.  I will  check out this build eagerly tomorrow morning  or later tonight after bedtime if i get too curious :)



Sent from my Samsung device


-------- Original message --------

**Comment by jzoss** [comment]
@teebee76 hope you feel better. Did you mean the latest build worked a lot better for you? That would be sweet. Anyway.. I was about to push the latest build from C.I. to an actual pre-release.. 1.5 beta 1.. I fixed a few more little bugs.

**Comment by teebee76** [comment]
I haven't tried the latest yet,  the version i have on my computer was the one from 2 days ago or something like that,  when we last "spoke" on github :)  your kids had the flu or something..  When my kid is tucked in I will fire up the computer to get the newest :)



Sent from my Samsung device


-------- Original message --------

**Comment by jzoss** [comment]
@teebee76 Kids just a cold :) I meant to ask the last version you tried worked better than the previous versions? That was the first to fix the resharper bug... I think :)

**Comment by PureKrome** [comment]
- What was the resharper bug, please?
- Can we ammend the title to: Release 1.5 discussion.
- Add a 1.5 milestone.

:balloon:

**Comment by jzoss** [comment]
@PureKrome -- The resharper bug was not one that we even had entered a bug.. but it had be noticed a by a few people.. Basically when using both the git plugin and resharper visual studio would get really slow.  There was a few causes to this, but I think the main cause I *THINK* was both plugins were using the on visual studio idle event to process ui state. The git plugin no longer uses it.. Also much of the work the git plugin was doing was blocking ui thread.. and I worked hard to clean all that up and do much of the work on background threads and only process on the UI thread when needed. Now.. this is one of those... crappy bugs.. that it's hard to know for sure if you have fixed.. because just because it's not happening as much now, does not mean its' 100%.. Hard to prove a negative.. 

Oh and version 1.5 .. that's only because the C.I. build really wanted to do the versions that way.. and I figured instead of fighting it, I'd just go with it...

**Comment by jzoss** [comment]
As promised .. Beta 1.. Might be a little rough around the edges.. but.. I think it's the best release yet.. It's fast, uses less memory, does not throw a bunch of exceptions :+1: So please break it :) tell me waht you think. 

Link -  https://goo.gl/hqh9ow

**Comment by PureKrome** [comment]
installed at now testing... :+1:

**Comment by teebee76** [comment]
Ooooh so excited right now!
 But at the moment my girlfriends using my computer,  and it's a quarter to midnight,  so I guess I'll just have to dream about the wonderful new world og git source control provider until I'm at work  tomorrow.
And I forgot to mention earlier that I too use ReSharper.  So the slowdown on my system was probably due to ReSharper not the structure of my solution/projects.

Let you know more tomorrow

Sent from my Samsung device


-------- Original message --------

**Comment by jzoss** [comment]
@teebee76 Tell me what you think.. Glad you are excited.. I have spent wayyyyyy to much time on this release especially for something I don't get paid :) But I admit it's fun. I work on mostly on internal software so I don't do much code that people actually see. Anyway @teebee76 and @PureKrome you guys helped a lot :) Thanks.. And to reward you.. Once I get a few minutes.. I'm going to ask for your opinion on a few things code wise .. I could use someone else ideas on a few things :) but now.. Fallout 4

**Comment by PureKrome** [comment]
any code reviews required? just post links to the GH code and we can review for you.

![](https://vice-images.vice.com/images/content-images/2015/05/07/i-ate-the-wasteland-being-a-cannibal-in-fallout-new-vegas-253-body-image-1431000629.jpg?resize=*:*&output-quality=75)

**Comment by jzoss** [comment]
You never know.. Might have lots of code for you to review.. Keep you busy.. Truth is I'm not exactly proud of the code.. It was is rough shape when I got it.. And while I think I've made it better.. It's still has that good old code smell. Lots of unused code that needs removing.. Fun times!

**Comment by teebee76** [comment]
Finally back at work with a little spare time to do some gscp-testing :) I notice the 2 latest builds failed, so I guess the last running is the 1.4.58 ? I'll have a look and report back with my findings...

**Comment by teebee76** [comment]
Somethings I've noticed are reported below, I made new issues for a couple of them, but the rest I'm not actually sure that are bugs, so I skipped the issue-creation until discussed.

In the pending changes window, the branch used to be in parentheses like this:
"Pending changes (develop)" 
Personally I like it better this way, because it makes it more distinct and visible, not part of the static title for the window that today will show: "Pending changes develop"

The pending changes window is now lightning fast!!! Really good job! Have not noticed any of the delay I did using an earlier version, no sign of slowdown after using it for a while either. 
... but it seems Solution Explorer glyphs are not updating, or does so very rarely. #25  

If I turn on the "Pending Changes Filter" in the Solution Explorer, it displays only the correct changed files, although the glyphs will on some of the files  show no changes. If I undo changes to a file, and save to the originally checked-in-state, it will not dissapear from the pending changes filter unless I turn it off and back on again... Not sure if the Solution Explorer was autoupdating before, because I was not aware of this "pending changes filter" until today. It sure would be nice if it did update automatically though.

Background-color has gone missing again... #24

Not sure if I remember correctly, but I think that if I doubleclicked a specific line of addition/deletion in the pending changes window the editor would open the file, and take me to this specific line of code? Not a biggie, but would be nice to have that back (if it was there in the first place, otherwise put it in the future-release-suggestion-box)

Looking forward to the next CI-build :)

**Comment by PureKrome** [comment]
Just some quick feedback. So far it feels way more stable/usable.

**Comment by jzoss** [comment]
@PureKrome Cool :+1:  That is what I am going for most of all :) Go to figure out a few things yet.. Like @teebee76 issues.. The Solution Explorer should update fast like the pending changes.. 

About the C.I. builds.. I was doing a bit of work on other branches.. trying things out.. I thought the C.I. build was just pointing the master.. but I guess it does it all :) Next Beta build will include of of these branches..It will speed up the  History UI that pops up... but I did it mostly to eliminated a ton of unnecessary code/projects..

**Comment by jzoss** [comment]
So.. I think I just found why the solution explorer is not updating.. And a word to the wise.. When setting up you test repos.. Add more then one file.. To test.. Because in my case.. The very first file in the repos update just fine.. Just no other file is updating.. Anyway if you want to see how fast it should be.. Change the first class in a project.. It's might fast.. Not that it makes it much better.. But hey at least I can fix it now..

**Comment by teebee76** [comment]
Great job finding that bug, just tested what you said, and not sure which file was determined to be the first, since this is a web-project, I can inform that would be the first .aspx-file in the root-folder.

The glyph changed like blitzed lightning just like the pending changes window, so it is almost scary-fast, how will I ever notice the changes when they are so fast? It's like the program knows which files I want to update :D 

anyway good job squashing that bug, the next build will be a real delight! maybe you'll even have time to fix that backgroundthingy ;)

**Comment by teebee76** [comment]
Noticed another bug while checking in files. Seems to me that when I check certain files, enter a comment and press commit half of the times the file will get staged instead of commited. 

![image](https://cloud.githubusercontent.com/assets/5390116/12257179/569babb8-b904-11e5-8afa-baca0f1c4926.png)

I checked in 4 files one-by-one the first committed, second was staged, third was committed (along with the already staged file) and the fourth was again staged. I don't know if this is random or that it actually alternates between committing and staging.

Did some more checking in with multiple files, and the pattern stayed roughly the same, first the commit went as it should then it just added the new files, on next commit on other files the new ones were staged, and then on the third commit they were finally committed. And now I am stuch with the last file that has staged status. 
![image](https://cloud.githubusercontent.com/assets/5390116/12257450/2f34ecb2-b907-11e5-9f58-de2185176772.png)
If I try committing it, I get this error.

![image](https://cloud.githubusercontent.com/assets/5390116/12257459/42369a36-b907-11e5-945e-bfb2338871de.png)

Which is somewhat funny message to get since the file is in fact staged for commit ... or so I thought ... 
I went to ye olde git bash and did a git status 

![image](https://cloud.githubusercontent.com/assets/5390116/12257492/81fc8bc6-b907-11e5-83ed-92decf5a96cb.png)

so it seems the pending changes window is out of sync somehow :/

I did a restart of VS and then the Pending Changes window ofcourse was "in sync" again

**Comment by jzoss** [comment]
Ok.. People .. Beta 2 is out.. should address many of the issues, but not all I guess :) Solution explorer is updating again :+1:  And the super important background  color is fixed. 	
goo.gl/V9eSlv

There is still a few things that need to be done.. If anyone has an idea on #23 I am looking for suggestions. I made one update .. I switched the code from .Net 4.5 to 4.5.2 because I read somewhere that may fix it :) Total shot in the dark. 

Anyway feedback would be cool.. It's Hoth cold today, so I may have time to work on it has I avoid the out doors tonight.

**Comment by jzoss** [comment]
Hello there.. Been sorta quite around here :) Anyway... I have just put up Beta 3 [https://goo.gl/Q7eRZM] It should fix #23, Also should fix a threading issue that could cause visual studio to slow down.. Anyway.. I am hoping this will be the last beta release.. @PureKrome + @teebee76  if you can hack on it hard.. Pay attention to building and add lot's of files.. If you or anyone else see times where VS pauses or slows down... mention it..

**Comment by PureKrome** [comment]
installed and testing over the next few days.

:clap: :+1:

**Comment by teebee76** [comment]
I will bring my computer home this weekend, maybe I will have time to use I don't know yet .. but will try . probably not going crosscontry-skiing if the temp stays below -15 anyway, so then I'll have time to do some coding instead :+1:  keep up the good work, will report anything suspicious

**Comment by jzoss** [comment]
@teebee76 @PureKrome Thanks.. Have a good one weekend.. @teebee76  Stay warm.. going to get cold.. wanted to go out this weekend as well.. but... -22c this weekend with -40c windchill .. Might just stay in myself.

**Comment by jzoss** [comment]
Hey guys.. A bit sleepy.. So much I didn't even make a release yet.. I want to clean the last few things up.. Anyway I will explain more later.. But I had to do a bit of a overhaul in a few places. .. Anyway need to clean up things are renaming and deleting files.. Anyway please install the release from ci.. @teebee76 and @Purekrome ... How's beta 3 going.. Been quite around here.. Either of you going to Build 2016?

**Comment by teebee76** [comment]
I will get the newest fra CI, and check it out. Haven't noticed anything particular with regular use of the version I am using now.. only the minor cosmetic things I mentioned earlier. 
When it comes to Build 2016, I hadn't heard of it until now.. And even if I had known about it, I still wouldn't have left. The company I work for probably wouldn't see the benefit and pay the travel-expenses, and it looks like it would be a 12+hour flight for me to get there... would have been fun though..

**Comment by teebee76** [comment]
ooh Nice, you fixed the parantheses around the branch :D

**Comment by PureKrome** [comment]
@jzoss so far (while i'm on holidays at the beach/pool) my small amout of coding == no noticed issues with the _previous_ Jan 15th release. :+1: :+1: So, so far so good IMO :smile: 

re: //build '16 - nope. I usually don't go to these things cause of the massive :moneybag: costs and I usually have a pretty good idea of what's coming as I keep my ear to the ground with shiny, new bits coming out of MS (for the area's I care about).

**Comment by teebee76** [comment]
did notice some out of sync action again :disappointed: 
and pending changes - refresh didn't help.

I noticed this kind of out-of-sync earlier at one point, but when I did a "git status" from the git bash, it was updated (might have been a coincidence?)

I also noticed that now all of a sudden the pending changes branch suddently no longer includes the parantheses that was there earlier today ... 

Solution Explorer Shows a Changed file. 
![image](https://cloud.githubusercontent.com/assets/5390116/12449604/b746687e-bf7e-11e5-99c4-8842e2c6e453.png)

Pending changes window shows 2 modifications, but no files ..... ? 
![image](https://cloud.githubusercontent.com/assets/5390116/12449623/d517c60e-bf7e-11e5-856f-33106c778f5e.png)

git status reveals the truth .. 
![image](https://cloud.githubusercontent.com/assets/5390116/12449640/f0d3674a-bf7e-11e5-80bd-e44e46af542a.png)

did a VS restart and ... 

![image](https://cloud.githubusercontent.com/assets/5390116/12449758/c26dd8bc-bf7f-11e5-85fa-a5be7ccb66fe.png)

Now it doesn't display my branch at all, it tells me there is 1 modification but at least it shows me the changed file :+1:

**Comment by jzoss** [comment]
@teebee76 You are right on a few counts.. First.. doing an external status can effect the status..So.. and I can go into much more detail later ... I have a file watcher .. it watches the git directory.. and it watches for 2 types of files events... File changes in your normal files.. like editing a code file and file changes in the .git directory of GIt events.. Anyway the git events tell me when a branch has changed or a commit, and even a status from a external tool.. I then can in theory update when evens happen.. Of course last night I was sleepy and committed out the GIT event to trouble shoot a different issues... and  forgot to put it back in :)

**Comment by teebee76** [comment]
crap I had written this down, but it was still in edit-mode when I left yesterday ... 

There is also something fishy happening when switching branches... 

starting off in my develop branch... everything looks clean. 
![image](https://cloud.githubusercontent.com/assets/5390116/12450512/8a26b5f0-bf84-11e5-8db2-c609c84de3e6.png)

![image](https://cloud.githubusercontent.com/assets/5390116/12450523/9acdb426-bf84-11e5-9887-994408d72906.png)

I switch over to my master branch... 
![image](https://cloud.githubusercontent.com/assets/5390116/12450533/b3585244-bf84-11e5-8021-e5b33eecab2c.png)

Standard updatemessage appears and I Reload All ofcourse ...
![image](https://cloud.githubusercontent.com/assets/5390116/12450549/d312975c-bf84-11e5-9606-ada09c80010f.png)

but .... 
![image](https://cloud.githubusercontent.com/assets/5390116/12450578/f71ed908-bf84-11e5-8d16-feaab7866088.png)
![image](https://cloud.githubusercontent.com/assets/5390116/12450596/08a2bb86-bf85-11e5-900c-d579b3af2c80.png)
Also in Solution Explorer
![image](https://cloud.githubusercontent.com/assets/5390116/12450637/2c778af0-bf85-11e5-8ffa-08627ac0f25d.png)
Select it in Pending Changes 
![image](https://cloud.githubusercontent.com/assets/5390116/12450650/43df4f84-bf85-11e5-8205-4f5b778616d9.png)
but no changes, so no changes to display .... 
![image](https://cloud.githubusercontent.com/assets/5390116/12450673/5d2082c4-bf85-11e5-8993-4a2bc76b1d32.png)
Again clicking Refresh does nothing ... 
.. but if I edit 1 file and save.. for instance the AssemblyInfo, mostly everything seems to be sorted out.. 
![image](https://cloud.githubusercontent.com/assets/5390116/12450732/c825e0aa-bf85-11e5-91d1-738a9b2deff1.png)
except it still doesn't show me the pending changes for that file ...

**Comment by teebee76** [comment]
Did some more testing today ... 

This time without invoking external git-events... I hope :smile: 

Starting off with a clean  working dir in branch develop.
![image](https://cloud.githubusercontent.com/assets/5390116/12476528/35d5ea56-c02a-11e5-96ad-4b7c2e081781.png)

Starting VS, first off I notice a difference from Solution Explorer and Pending Changes...
![image](https://cloud.githubusercontent.com/assets/5390116/12476579/a021dd3e-c02a-11e5-8bc7-c749d3d2fd10.png)

SE displays the branch I am on, Pending Changes does not.. 

next I change my version in AssemblyInfo and save the changes.
Looks great, Pending Changes displays the file and the changes inside. (still no branchinfo in the header) 

I change the branch to master, but no identification that I have switched to master is visible
![image](https://cloud.githubusercontent.com/assets/5390116/12476834/44932c82-c02c-11e5-9058-944746a6bcca.png)
I do see that if I click the switch-branch my current branch is preselected
![image](https://cloud.githubusercontent.com/assets/5390116/12476872/9063de9a-c02c-11e5-8ecc-cd1e8437e9ec.png)
I switch back to develop and commit the changes.
Now the pending changes displays my branch .. 
![image](https://cloud.githubusercontent.com/assets/5390116/12476934/d770d9e6-c02c-11e5-8201-fa879ed925be.png)

I switch back to master and get the updated outside vs warning .. choosing yes to all 

I should now be on master branch with no modified files ....
![image](https://cloud.githubusercontent.com/assets/5390116/12476986/211e8836-c02d-11e5-8396-e19015df6a02.png)
Solution Explorer still says (develop)
Solution Explorer shows changes to AssemblyInfo
Pending Changes says develop
Changed files status says 1 modified file 
Changed files doesn't show any changed files (as it should) 

I edit my version numer once more and save the changes. The Pending changes now shows the file (still displaying the wrong branch in the header) 
![image](https://cloud.githubusercontent.com/assets/5390116/12477072/a6dbbe6c-c02d-11e5-9f2b-6a919189a8b4.png)

I change it back to what it was, and saves the file once more.
still showing me the same pending changes..
Clicking refresh -> nothing happens
Selecting another random file, and does one change and saves.
Pending changes are updated. The Assemblyinfo file dissappear, and the new one appears. 
however  Solution Explorer is not updated 
![image](https://cloud.githubusercontent.com/assets/5390116/12477169/5d4ab87e-c02e-11e5-89eb-c4ef796e3b4c.png)
I remove the change I did and saves yet again, should now be back to the clean state in master, however.. 
Solution Explorer display (develop) and changes to AssemblyInfo
Pending Changes display develop and changes to Antallbegrenset.aspx 

I switch back to develop get the same warning select update all.
and no changes to SE or PC. 

I finally go outside to git bash to check my stat, and 
![image](https://cloud.githubusercontent.com/assets/5390116/12477323/2c062158-c02f-11e5-99b5-a77a2957b5b0.png)

**Comment by jzoss** [comment]
Ok.. So... ever have one of those little bugs.. that all the sudden turn into a big PITA :) Where you redo annoying amount of your code to fix.. Well.. I found one :) I may go into a huge amount of detail later.. Once my brain recovers... But everyone get excited for Beta 4 https://goo.gl/8nyt29  @teebee76 this should solve the things you were running into.. @PureKrome .. You should sit by the pool.. and be lazy.. but if you want you can test it a bit.. 

Few things to finish up.. mostly around loading and unloading projects.. but I should be done with that pretty quick as long as no one finds more bugs ...

**Comment by PureKrome** [comment]
Just leaving in 10 mins on a road trip for the weekend :+1: will try this in a few days when i get back.

Keep up the work/posts @jzoss :cake:

**Comment by teebee76** [comment]
Super ! 
just installing it now and will hack at it during work-hours :smiley: 

I really like your dedication @jzoss  :+1:

**Comment by jzoss** [comment]
Ok.. I'm a bit excited :) I finally think I have it done it.. After way too much time spent.. and learning way too much about internal threading of visual studio .. Tonight I think I finally fixed the last nagging issue... And so... tonight I give you 1.5 RC 1.. I will hopefully become the final release.. though I may fix #29. Anyway.. this all the internal stuff should be done.. Please try switching branches a lot.. make sure you don't see any blocking ui, or slowdowns.. Anyway here you go.. https://goo.gl/eVtgBz

**Comment by teebee76** [comment]
Ooh looking forward to testing it today :+1:

Did a quick branch switch before I got started, and must sadly report that it doesn't change like before :disappointed: 

Opening visual studio I have 2 files with changes on my develop branch
![image](https://cloud.githubusercontent.com/assets/5390116/12574902/5b677f78-c408-11e5-87f4-8a5e5e3fe5be.png)
I then switch to my master without staging those changes, and VS window quickly updates, but no visual indication that I am actually on another branch.
![image](https://cloud.githubusercontent.com/assets/5390116/12574933/88e0c9b4-c408-11e5-93d5-3f2aef28efb4.png)
So I do a git status from git bash 
![image](https://cloud.githubusercontent.com/assets/5390116/12574948/b4c00d24-c408-11e5-9696-3c5c3f8fdfc7.png)

(sidenote: I did all of this before closing VS and doing the process over to grab the screenshots, and the first time around I believe I saw parantheses appear in det pending changes window heading, could be mistaken, but pretty sure I saw them)

Will continue testing different things while coding today, and will summarice at the end of the day if I find anything aside from this.

**Comment by teebee76** [comment]
by the way, the new Icon on the extension, there's nothing wrong with that. It's sharp and easily recognizable for people using GIT from before, it looks pretty much like the original icon for GIT :smile:

**Comment by 0x084E** [comment]
I've played a bit with branch switching and noticed a little problem:
If a project or solution file changes and you force VS to show the dialog which offers you to reload the project/solution (e.g. with klicking into the editor) while the provider is still busy with switching, the dialog is shown and immediately hidden (i think in the moment the provider is finished with switching). After this you are not able to use VS anymore because it behaves tßlike a dialog is still open.

I hope this was halfway understandable. Here the steps to reproduce:
1. Open a solution, I think the larger the better, switching the branch should take some time
2. Switch the branch. Make sure the solution file changes between the branches.
3. Immediately click anywhere into VS to force the 'reload-dialog' to appear.
4. Wait until the dialog closes automatically
5. You are now locked out. Only solution I've found is killing VS

Let me know if you need additional information.
Btw.: If you avoid that the 'reload-dialog' appears, branch switching is working well!

**Comment by teebee76** [comment]
@0x084E when you switched branches, did the it visually change for you? Maybe it's just my structure or setup that prevents my VS to display the correct branchname in the title of Solution Explorer and Pending Changes windows ...

**Comment by 0x084E** [comment]
No, it doesn't. I didn't notice this because of the solution file change. If I reload the solution after switching the title is fine, of course.

**Comment by teebee76** [comment]
A short update. Noticed I had one update of Visual studio pending .... so I did that update (which took like 30minutes or so .. )

started my project again and voila
![image](https://cloud.githubusercontent.com/assets/5390116/12580598/26cf873c-c431-11e5-9353-9f34ff5bd0ce.png)

I thought I'd mention what update I did, so I went over to Help About to find versionnumber etc.. 
![image](https://cloud.githubusercontent.com/assets/5390116/12580622/4f04992c-c431-11e5-9229-0aa8ee5d0069.png)

I then also noticed a mismatch in version numbers for the plugin ...
![image](https://cloud.githubusercontent.com/assets/5390116/12580661/97ed5868-c431-11e5-82f6-9084c30e5d44.png)

although the right version is displayed under Extensions and updates 
![image](https://cloud.githubusercontent.com/assets/5390116/12580678/bdba64aa-c431-11e5-8afa-16138a2e990c.png)

It sure isn't important in any way, I just thought I'd mention it :smile:

**Comment by jzoss** [comment]
Ok Everyone.. and of course @teebee76 .. Do you want ( ) around all the branch names? 

I think I know why the branch names are not changing.. the answer is they are.. bu then they are getting reset :)  

@0x084E Thanks for the heads up. I hope I can find this issue.. I hope It's fixed by using the new threading fix I did.. Just a note.. I THINK if you wait long enough visual studio will come back.. But the freeze used to happen ever you changed a lot of files. So.. it's gotten a bit better.

**Comment by 0x084E** [comment]
I've tried this a little bit more and I think I've found another detail:
The problem described above only happens if you click into VS while the 'Switch Branch'-Dialog is open (but after you clicked 'OK'). So it seems to me that closing the 'Switch Branch'-Dialog also closes the other dialog in any way.
After this happened, if you click into VS you hear the typical sound windows is playing if a modal dialog is still open and you click into the window below. I've waited for minutes and VS will definitively not come back.

**Comment by jzoss** [comment]
@0x084E -- Interesting.. I take it you are talking about the switch branch in the pending changes window?

**Comment by 0x084E** [comment]
@jzoss Yes, I am. Sorry if this wasn't clear.
I had a look at the source code and tried changing the btnOK_Click method in BranchPcker.xaml.cs to be synchronous (with accessing Task.Result instead of using await Task) and now the switch dialog never gets closed (but the branch gets switched).
I'm neither a VS Extension nor a WPF expert so maybe I'm missing something obvious but this seems a little bit weird to me.

**Comment by jzoss** [comment]
@0x084E Ha .. I was not either, before I went on this crazyness.. First do 1 test for me.. Do a switch through an external tool .. TortoiseGit is what I like :) and make sure it does not freeze for you.. You are correct about the code .. you found the right line.. Give me 15 minutes .. I was to see if I can you a C.I. build to test..

**Comment by 0x084E** [comment]
Switching the branch with TortoiseGit (which is my favorite, too ;)) is working well (even with TortoiseGit Toolbar extension inside VS).
Also the switch dialog is working if  the 'solution has been modified outside of VS' dialog is not popping up between `await repository.CheckoutAsync((GitBranchInfo) comboBranches.SelectedItem)` and `window.DialogResult = true`

**Comment by 0x084E** [comment]
@jzoss I think I know what is going on, it is a problem with the WPF Window.ShowDialog method,
According to the <a href="http://referencesource.microsoft.com/#PresentationFramework/src/Framework/System/Windows/Window.cs,cebef344eecbe702,references" target="_blank">Source</a> this method remembers the active window and restores it aufter closing. Now the following happens:

1. Show switch dialog, select branch, click OK
2. The switching starts which causes the solution file to change, the switch dialog is waiting for switch to finish
3. Click into VS, the ''solution has been modified outside of VS' dialog appears as modal dialog
4. switch finishes and the switch dialog is closed which causes the restore of the main window as active window

What you have now is that the 'modified' dialog is send to background while being modal which causes some kind of 'ui deadlock' (you can't acces the dialog because its behind the main VS window and you can't access the main VS window because it is waiting for the dialog)

If I'm right, one solution would be closing the window before starting the switch, but I don't know if this causes other problems (e.g. when showing a MessageBox in the cause of an error).

What do you think?

**Comment by jzoss** [comment]
@0x084E Once again you are correct.. First ... Sweet that the other switch works..TGIT rocks... i just use the git command through the right click in the solution explorer .. Anyway .. there is a new C.I. build.. I put in most of the branch fix .. .I think.. I took the switch off the popup .. put it on a background thread.. I am checking it in right now .. hold on..

**Comment by jzoss** [comment]
@0x084E Here is the C.I. build .. can you still break it? https://ci.appveyor.com/project/jzoss/git-source-control-provider/build/artifacts

**Comment by 0x084E** [comment]
No way, I really tried, but it works :) :+1:
I've also tried creating a new branch with this dialog which also works very well. A little thing I've noticed while trying this is that the branch dropdown list is not refreshed and doesn't contain the created branch when reopening it.

No not really, it lacks multi repository support and the 'open bash' command only works with msysgit not the new git-for-windows. I'm not using it much.

**Comment by jzoss** [comment]
@0x084E  Cool.. and just to be difficult.. New C.I. build.. Branch name fixed changing when you switch should be fixed.. and I just fixed the updated branch list thing. Man.. I'm fast today.. or avoiding my  real work :+1:

**Comment by jzoss** [comment]
@teebee76 .. You just had to tell me about the version thing.. didn't you :) 
![version](https://cloud.githubusercontent.com/assets/3586254/12596680/a7d35808-c446-11e5-81de-566ec4d821c8.PNG)

**Comment by jzoss** [comment]
Hi all... I did not expect to do a RC 2 release so quickly, but I had a lot of help today. .. so here you go.. https://goo.gl/7OSuf7 

So.. now keep hacking at it.. @PureKrome Make sure it runs as well as before... @teebee76 and @0x084E  you seem like you have a talent at finding branching issues.. Anyway thanks for your help.. All I plan to do not is update the docs ... fix any bugs.. I was hoping to wait to push the final release until libgit2sharp .22 was released... but not sure when that is going to happen.. 

Also I was in a good mood and put in #29 .. It you guys could test that too.. 

Thanks for your help everyone.

**Comment by teebee76** [comment]
Wow a lot of activity last night :+1: 

@jzoss  ( ) or not doesn't really matter, as long as it is consistent. The way it is now SE always has it and PC has it sometimes..  Personally I think it looks better encapsulated in parentheses since it is not part of the original title, but indicating a state that might change.

...and @jzoss when it comes to the version number, you can always count on me to point out the most irrelevant things :smile: 

will get the latest from CI and do some branching yet again

**Comment by jzoss** [comment]
Warning brain dump coming. Sorry to everyone it's long :) except @PureKrome I hope is detailed enough. 

## tldr version –
Threading is hard! It’s even harder when you are calling COM components .. You need to be careful and follow the rules

## Long Version 

## Background
So, I noticed in 1.4.2 in some cases it could freeze Visual Studio. This was unacceptable to me, but I had no idea where the fix would take me. So the goal for this release was initially to increase performance and reduce memory usage, but as I discovered how easy it was to crash VS with bad code, I knew I had to solve that before the release. On the code side, I wanted to start reducing coupling and separate the pending changes and solution explorer integration. The idea behind that is MAYBE eventually, in a later release, have the pending changes work without the SCC part. 

So a bit of backstory, before I forked this code, it worked on a loop. Every few seconds it would refresh all the glyphs and the pending changes. This worked…. Ok … but it could cause slowdowns.. Especially in larger projects. It was simple which was good, but didn’t work all that well in many cases. 
In the first releases after I forked the code, this did not change, until 1.4 when I stared to update the code. It’s not that the code was bad when it was written, but the original plugin was for vs2008 – 2013 so it had been around for a while, based on .net 3.5. 
##The Problem 
So in fixing performance, first this I did was switch all the git command to use lib2gitsharp. This improved performance a significant amount. It was still using the infinite loop to update the glyphs and pending changes. Next I added a FileSystemWatcher to the GitRepository class. The idea is why not monitor the code folder. It threw an event whenever a file was changed. Performance was increased, however as I increased performance. Locking and slowdowns become worse. 
So.. bit of context for this next part.. There are 4 big components you need to know about to understand.. Anyway if this makes sense you know ow a big part of this extension works. 

### SccProviderService
This is the service class that, well it still does a bit too much, but basically it updates the glyphs you see in the solution explorer.. also handles right click menus and all that.. But the update glyph status is what’s important here.

### Pending  Changes 
It’s the pending changes window, that also shows you what files have changed.  

### GitRepository
This is a class that uses liib2gitsharp and a filewatcher to detect changes that happen in a repo. It is also used to do any git command, like commit. The most frequently used git command is the RetrieveStatus command.  As you might imagine, we get the status a lot. Anyway, it has been almost completely rewritten since 1.4
### RepositoryManager 
 This is a singleton class.. And it’s the glue that holds the parts together and allows for multiple repos. It is completely new class since the fork. So it maintains a collection of all the GitRepos in a solution, as well as what repo is the “main” repo for the solution and “active” repo. It listens for events from the repos and forwards them on to the SCCService and pending changes window.

So it goes like this. The Pending Changes window gets the ActiveTracker from the RepositoryManager(RM) and listens for a GitStatusUpdate event. When event happens it simply refreshes the changed files.. The SccService listens to changes from all the repo (by connecting to one event in the RM that listens to all know repos) and when it get one. It finds the files that have changed and updating the glyphs.

So using the events, with async handlers made everything really fast.. But lead to lots of pausing a crashing.. So. I started at the GitRepo class and started to make things more efficient to try and fix the crashing. I did a lot of things.. Like the GItRepos used to just send the filename that changed, and I changed that to also send the status of all the files that changed (I can go in to the reason for all that at some point if anyone cares) I made the SccService only change glyphs if it thinks the status has changed.. I made every method async through the update cycle, and little by little it got better.. but still would freeze and block.

## The Fix 
So. It took me forever to get this far.. and the fix was not really that hard. I wish I would have figured it out before, but I could not find the problem.. So at this point I had made the code pretty efficient, worked around the issues I had with libgit2sharp, and suddenly it was VERY clear I was having a issues with multiple threads. VS would freeze… I could pause the debugger and see thread contention.  
So.. I found this blog https://blogs.msdn.microsoft.com/andrewarnottms/2014/05/07/asynchronous-and-multithreaded-programming-within-vs-using-the-joinabletaskfactory/
And the nuget package Microsoft.VisualStudio.Threading.Analyzers (https://www.nuget.org/packages/Microsoft.VisualStudio.Threading.Analyzers/)
It talks about new threading in VS2013 and 2015.. It is much simpler than previous versions.  I was on the right track.. The analyzers package showed me all my threading issues at first.. and that was awesome. 
So what I learned 
1.	Uses Async/Await – You will be on a background thread, and that is good.. but you need to be aware of all the next rules. 
2.	NO async void methods! Unless it is an event handle. If you have an exception in a async void.. you are going to block.. they should be async Task  
3.	If you do anything with a COM object YOU MUST BE ON THE UI THREAD. I know that I did not guarding against this so sometimes the crashes happened because I was on the background thread. It includes casting the object… So with the new threading before you do anything with a COM object ..
```c#
	await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();

	//Later do some work on a background thread
	await TaskScheduler.Default;
 ```
Note: this also means that is made sense to cache any info from COM objects, for instance what files go to what projects. To get this info you need access to the COM object and to access the COM object you have to be on the UI thread. 

So following the new rules in the blog and the nuget analyzer, I was able to fix a lot of crashes.. Visual studio stopped crashing.. but it still paused.. It would recover.. Sometimes after a long while. Then I discovered the secret with The SiwtchToMainThreadAsync(); 

### Rule 3 - Super Important – Run UI threads with the correct background priority
If you use await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(); you are taking the UI thread at the highest priority. WHICH means you are even blocking user input. So.. the last big slowdown that happened was when you were doing a branch with lots of files changes.. and I tracked it down to waiting for the glyphs to update.. 
So. Not covered in the blog is an, not all ui threads are treated equal.. I want to run something on the U.I. thread, but not block the user.. Then I need to start a async operation a little

```c#
                await ThreadHelper.JoinableTaskFactory.RunAsync(
                    VsTaskRunContext.UIThreadBackgroundPriority,
                    async delegate
                    {
                    // On caller's thread. Switch to main thread (if we're not already there).
                    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
                    // Now on UI thread via background priority.
                    await Task.Yield();
                        // Resumed on UI thread, also via background priority.
                    }); 

```

So my update glyph went from 
```c#
	await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();

	foreach (var project in projects)
              {
                  if (project != null)
                  {
                      project.SccGlyphChanged(0, null, null, null);
                  }
                  await Task.Yield();
              } 
```

To 

```c#
	await ThreadHelper.JoinableTaskFactory.RunAsync(VsTaskRunContext.UIThreadBackgroundPriority,
               async delegate
           {
             // On caller's thread. Switch to main thread (if we're not already there).
             await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
 
             // Now on UI thread via background priority.
             await Task.Yield();
             foreach (var project in projects)
             {
                 if (project != null)
                 {
                     project.SccGlyphChanged(0, null, null, null);
                 }
                 await Task.Yield();
             }
             // Resumed on UI thread, also via background priority.
               });
 ```


Anyway..I have found that in almost all cases you should be on a background UI thread, or a background thread.

**Comment by PureKrome** [comment]
Awesome post @jzoss :smile: :clap: :clap: 

The most useful piece of that post was your learning experience _and_ the steps you took to help resolve the problem. For me, this is gold. So thank you :)

First rule of `async/await`: `public async Task (or Task<Foo>) FooAsync(..)`... never return `void`. (well 99.999999999999% of the time).

And finally ... [this is all of our first attempt at async programming](https://twitter.com/jonathansampson/status/676487374495342592)

**Comment by teebee76** [comment]
Excellent and educational post @jzoss! I read it last night on my mobile phone, even though it is quite long I think I will read it again today on my computer. 
Very insightful, and surely will help me understand the though behind the complete process of the plugin, if I should ever get the source and do some experimenting myself :) Haven't done much async and await myself, but I heard it is a less messy kind of multithreading than we had to do before.. Maybe I will need to come up with an internal project so I can dive into it myself. Lots of cool and fun stuff has come along with the latest versions of .net and c#, so gotta keep up :D 

If I haven't got the latest version, I will get it once more today, and hopefully do less SQL stuff today, and more C#

**Comment by 0x084E** [comment]
Interesting post, thanks for sharing! Also an interesting usage of async/await for switching threads, I'm already thinking where I can build / use somthing similar in my programs :smile:

Btw.: RC2 is running fine on my machine, no issues found so far :+1:

**Comment by jzoss** [comment]
@PureKrome , @teebee76 @0x084E - Thanks guys.. Good to know I'm not just talking to myself :) I guess I could have gone into even more detail! Anyway.. I may clean up a few more little things.. mostly making sure thing that might throw an exception are handled.. remove a few bits of dead code blocks.. Update readme ..May have 1 final RC... What do you guy think.. is it ready to be released.. My big issue with this... I don't really have any automated tests.. I have not found a good way to test it all yet.

**Comment by teebee76** [comment]
I noticed something earlier today, but didn't manage to reproduce it, so I didn't post it earlier... but I will mention it now anyway in case others see it too. (and because you just reminded me of it now)

I was working on my other solution today, the on that doesn't have the entire solution under the same repository. and have some projects under individual repositories, and some that are not source controlled at all.. While on this solution I wanted to test the unit testing using Xunit.. and like all the others out there I created a new project for the tests and linked it to the main project. 

I then ... (don't exactly know where in the proccess, if it was when creating the project, or getting the nuget-packages for it or whatever ... ) I suddently noticed that it was marked in Solution Explorer as checked into source control..  As was alle the other projects that was not under any repository. 
I double checked to make sure I didn't create the new project accidentally under an existing git-repository, but when in git-bash it told me not under source control. 

So some kind of hickup. I restarted VS and everything was back to normal. 

Deleted the project and created a new one to try reproduce without success .. 

maybe I will try again tomorrow with som more nuget package stuff etc ..

**Comment by jzoss** [comment]
@teebee76 .. OK there you go again :) Finding bugs.. OK check out the latest C.I. build. I was going to make it a RC build.. But then I forgot.. Anyway guys if you'd could try it out. So @PureKrome , @teebee76 , @0x084E .. Opinion.. The latest ci build.. Release or let people test it for a while longer?

**Comment by PureKrome** [comment]
I've been using the latest from a day or so ago .. since a day. gimme a week of slamming it. so IMO: wait.

**Comment by jzoss** [comment]
@PureKrome  You are wise like 
![yoda2](https://cloud.githubusercontent.com/assets/3586254/12680370/15546f1e-c66f-11e5-88a9-5df21d9734a1.jpg)

OK people .. take the latest C.I. build and hack away :)

**Comment by jbestrom** [comment]
I haven't been using it a ton because of VS/Resharper issues but from what I have used it I think things are good, then again I don't do a lot of branch switching so I can't say on that but all the pending window changes are great.

**Comment by jzoss** [comment]
@jbestrom .. Is this Resharper issues with the plugin? or Unrelated?

**Comment by jbestrom** [comment]
Unrelated to the plugin. Resharper is trying to index generated files when it isn't supposed to so it makes VS grind to a halt. With resharper off this plugin still works as expected.

**Comment by jzoss** [comment]
@jbestrom I noticed it did a lot of strange things when I was debugging. Make sure you turn off solution wide analysis .. But yea.. Resharper is greedy with threads :)

**Comment by jzoss** [comment]
Hey all.. Just a FYI.. I made a new RC build it the release tab. It's the same as the latest C.I. from 4 days ago, but a little easier to find.. Maybe get a few more testers out there.. Anyway, once I release, I 'm just going to update the version on this build and push.. it should auto-update for everyone. Anyway as long as no one else find more bugs ( @teebee76 !!!  ) and @PureKrome says it's cool (He is wise like Yoda and has a bunch more twitter followers than I do :) ) Then I'll probably release this week or early next week.

**Comment by PureKrome** [comment]
I've been using it pretty much since the latest drop was put up (as mentioned above).

So far - seems to be working for me :+1: 

I also switch branches _a lot_ (mostly from the Explorer, not inside VS) and things are working fine :smile: 

So far so good @jzoss !

**Comment by jzoss** [comment]
@PureKrome Cool.. That's how I generally use it as well.. I commit sync from the solution explorer ... anything else I do I use TGIT.. 

I may actually have to release this.. The one thing that is clear, IMHO, is that it is better and more stable than 1.4.2

**Comment by jbestrom** [comment]
If this was a companies product you would have been forced to ship it 2 weeks ago :)

**Comment by 0x084E** [comment]
... from people who don't know what software development actually is :)

I think it is ready for shipping. I haven't noticed any problems the last week and as @jzoss said it's definitely better than 1.4.2

**Comment by PureKrome** [comment]
:shipit:

**Comment by jzoss** [comment]
@PureKrome @teebee76 @0x084E @jbestrom 

It is done.. 1.5 is in the release tab.. also uploaded to he gallery .. It should auto-update, or download it manually.  Thanks everyone .. I could not have done it without you help!! If anyone wants to help a little more.. you could tell your friends...or mention he plugin on twitter or something, I would greatly appreciate it.

**Comment by PureKrome** [comment]
Good work @jzoss and friends :+1: :cake: :balloon: 

From here, i'm interested in seeing if this will be merged up to upstream ... but that's a discussion in the other issue....

**Comment by jzoss** [comment]
@PureKrome Do you mean #13 ? I'm not 100% sure.. I go back and forth on that a bit. I am not sure what the best path forward is :) But as you said.. it's another issue.

**Comment by teebee76** [comment]
@jzoss great work, I really appreciate the time and effort you've invested in this project. 
First of all because it made me able to use my favourite extension after upgrading to vs2105, and ofcourse the additional bonus of making it superfast, and now also bug free (fingers crossed).

When it comes to the bugs, I haven't noticed anything, but then again I've not been able to test much either the last week :/ will continue to follow the project here on github, and will continue to report any irregularities I notice.

When you've made your company happy for doing some work for them again, and feel in the mood on expanding with new feautures I'll bring in some suggestions we can discuss ;)  Allthough I guess that depends on this project getting  merged or not.

I'll be sure to demand everyone I know to download this extension ;)

**Comment by jzoss** [comment]
@teebee76 Yes new features! That is more fun than merging.. anyway While I may take a few days off from this open source project :) I don't want to leave it how it is now. We do need to figure out the merging thing at some point.. but we can talk about that else where.. So.. though it makes me sad sorta nostalgic way... It is time to close this thread. 1.5 is out, and once again.. thanks everyone for your help and support.. I would have probably never done this with out you guys. I actually looked forward to status updates from you people. So @PureKrome @teebee76 @0x084E @jbestrom and anyone else please stop by the new release thread #30 .. I will post a update on merging there later.. but I love to hear where people think we should go next.
